engine/src/main/java/org/pentaho/di/trans/steps/csvinput/CsvInput.java

```diff
@@ -0,0 +1,1179 @@
+
+      if ( data.filenames == null ) {
+        // We're expecting the list of filenames from the previous step(s)...
+        //
+        getFilenamesFromPreviousSteps();
+      }
+
+      // We only run in parallel if we have at least one file to process
+      // AND if we have more than one step copy running...
+      //
+      data.parallel = meta.isRunningInParallel() && data.totalNumberOfSteps > 1;
+
+      // The conversion logic for when the lazy conversion is turned of is simple:
+      // Pretend it's a lazy conversion object anyway and get the native type during conversion.
+      //
+      data.convertRowMeta = data.outputRowMeta.clone();
+      for ( ValueMetaInterface valueMeta : data.convertRowMeta.getValueMetaList() ) {
+        valueMeta.setStorageType( ValueMetaInterface.STORAGE_TYPE_BINARY_STRING );
+      }
+
+      // Calculate the indexes for the filename and row number fields
+      //
+      data.filenameFieldIndex = -1;
+      if ( !Utils.isEmpty( meta.getFilenameField() ) && meta.isIncludingFilename() ) {
+        data.filenameFieldIndex = meta.getInputFields().length;
+      }
+
+      data.rownumFieldIndex = -1;
+      if ( !Utils.isEmpty( meta.getRowNumField() ) ) {
+        data.rownumFieldIndex = meta.getInputFields().length;
+        if ( data.filenameFieldIndex >= 0 ) {
+          data.rownumFieldIndex++;
+        }
+      }
+
+      // Now handle the parallel reading aspect: determine total of all the file sizes
+      // Then skip to the appropriate file and location in the file to start reading...
+      // Also skip to right after the first newline
+      //
+      if ( data.parallel ) {
+        prepareToRunInParallel();
+      }
+
+      // Open the next file...
+      //
+      if ( !openNextFile() ) {
+        setOutputDone();
+        return false; // nothing to see here, move along...
+      }
+    }
+
+    // If we are running in parallel, make sure we don't read too much in this step copy...
+    //
+    if ( data.parallel ) {
+      if ( data.totalBytesRead >= data.blockToRead ) {
+        setOutputDone(); // stop reading
+        return false;
+      }
+    }
+
+    try {
+      Object[] outputRowData = readOneRow( false, false ); // get row, set busy!
+      // no more input to be expected...
+      if ( outputRowData == null ) {
+        if ( openNextFile() ) {
+          return true; // try again on the next loop...
+        } else {
+          setOutputDone(); // last file, end here
+          return false;
+        }
+      } else {
+        putRow( data.outputRowMeta, outputRowData ); // copy row to possible alternate rowset(s).
+        if ( checkFeedback( getLinesInput() ) ) {
+          if ( log.isBasic() ) {
+            logBasic( BaseMessages.getString( PKG, "CsvInput.Log.LineNumber", Long.toString( getLinesInput() ) ) );
+          }
+        }
+      }
+    } catch ( KettleConversionException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)